lrathoni/Lullaby-for-a-fish
# Added, improved interactive objects

### What

#### Improvements 

* The *FlutePlayer* behaviour now uses Unity's State Machine implementation. It is not as now animated.

* Reworked interactive object components. They are now more fragmented and stackable. For instance, you can add a glow interaction on top of a hover.
* Reworked pause menu trigger, now invokes an event.

#### Fixes

* The hover interaction now actually makes the object hover and not just go to infinity and beyond.
* The flute is now muted upon entering the pause menu.

#### New features

* Added proper conditions for triggering OnNote events with *InteractiveRangeModifiers*, depending on the view of the camera and the distance to the object.
* Added a *PlayerAttracted* interaction which causes an object to be attracted to a position relative to the player.

#### Known issues

* ~~Flute playing is now cluckier, you can only press a single note at all times. This is done so the flute can be properly animated.~~ This was fixed.
* The flute and hand no longer move. They used to be animated by script, and should now use Unity's Animator.

### How to test

1. Check out this branch.
2. Open the prototype Hub scene in Unity 2019.4.
3. Enter Play mode.
4. Play the flute as usual.
5. Enter pause menu while playing and notice that the flute is muted.
6. Leave pause menu.
7. Play around with the objects. By this point anything that behaves strangely can be considered a bug, let me know if that's the case.
8. **Finally, let me know if any interaction should be added or changed to accommodate a puzzle or game mechanic!**

### Screenshots

* What I get while playing the flute. The Flute animator view should be active if the *UserInterface* game object is selected in the inspector.

![screenshot](https://user-images.githubusercontent.com/44610878/104793858-7985d880-57a4-11eb-95a9-dda38b73706b.png)
